File: solutions/clojure/kindergarten-garden/1/src/kindergarten_garden.clj
(ns kindergarten-garden)
(require '[clojure.string :as str])

(def plants {\G :grass \C :clover \R :radishes \V :violets})
(def students {0 :alice 1 :bob 2 :charlie 3 :david 4 :eve 5 :fred 6 :ginny 7 :harriet 8 :ileana 9 :joseph 10 :kincaid 11 :larry})

(defn partitioned [diagram]
  (map #(partition 2 %)(map seq (str/split diagram #"\n"))))

(defn garden [diagram]
  (apply merge-with into
    (for [row (partitioned diagram)]
      (into (hash-map) (for [[k v] (map-indexed vector row)]
                   [(students k) (into [] (map #(plants %) v))])))))

File: solutions/kotlin/kindergarten-garden/1/src/main/kotlin/KindergartenGarden.kt
class KindergartenGarden(private val diagram: String) {

    private val plants = mapOf('C' to "clover", 'G' to "grass", 'R' to "radishes", 'V' to "violets")

    private fun getPlant(code: Char) : String {
        return plants.get(code) ?: ""        
    }

    fun getPlantsOfStudent(student: String): List<String> {  
        val idx = student[0] - 'A'
        val offset = diagram.length/2 + 1 + 2*idx
        return listOf(getPlant(diagram[2*idx]),
                      getPlant(diagram[2*idx+1]),
                      getPlant(diagram[offset]),
                      getPlant(diagram[offset+1]))
    }
}
